### DISCUSSION TITLE
Dependency override not working

### DISCUSSION THREAD

**Discussion by chrobotm** [fastapi/fastapi#6245] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter, Depends
from ..dependencies.get_user import get_user

router = APIRouter()

@router.post("")
async def my_func(user: str = Depends(get_user)):
     ...
    return {...}
```

### Description

I'm trying to run some tests with pytest against the API but my dependency override is not working. It still tries to run the function as if there was no override. All other tests where we don't need to override a dependency work fine 

`tf/python/api/dependencies/get_user.py`

``` python
from fastapi import Header
from typing import Union


def get_user(xyz: Union[str, None] = Header(default=...)):
    return ...
```

`tf/python/api/endpoints/myfile.py`

``` python
from fastapi import APIRouter, Depends
from ..dependencies.get_user import get_user

router = APIRouter()

@router.post("")
async def my_func(user: str = Depends(get_user)):
     ...
    return {...}
```

`tf/python/api/tests/test_myfile.py`

``` python
import responses

from fastapi.testclient import TestClient
from ...main import app
from ..dependencies.get_user import get_user

def override_get_user():
    return "asdf"

app.dependency_overrides[get_user] = override_get_user

client = TestClient(app)

def test_my_func():
    resp = client.post("/api/...", headers={"xyz": "dummy"})
    assert resp.status_code == 200
    assert resp.json() == {...}
```

### Operating System

macOS

### FastAPI Version

0.78.0

### Python Version

3.9.13

**Comment by jonatasoli** [comment]
You need use `dependency_overrides` before declare `TestClient`.

**Comment by chrobotm** [comment]
Thanks, I've just tried that but it didn't work either. I've updated the main post with your suggestion

**Comment by chrobotm** [comment]
in `tf/python/api/tests/test_myfile.py`, importing:

``` python
from api.dependencies.get_user import get_user
```

instead of:

``` python
from ..dependencies.get_user import get_user
```

seems to have done the trick

**Reply by dctaf** [reply]
This fixed my problem. This should be mentioned in the documentation.

**Comment by ekivolowitz** [comment]
@chrobotm thank you for putting an end to two days of debugging dependency injections. Just needed to find the right comment :(

**Comment by netanel-haber** [comment, marked as answer]
If anyone has more trouble with this: Mounted sub apps using e.g.:
 ```python 
app.mount(path, subapp)
```
won't inherit `dependency_overrides` and will need to have the overrides registered again manually.

**Comment by SayeedKhan21** [comment]
hey everyone  I'm having a similar issue where I am trying to test reading a user at  /users endpoint ,however the app seems to read from real db rather than test db , the override_get_db doesn't seem to run . Any suggestions ?

**Reply by rubywwwilde** [reply]
I believe the issue is that you are importing your dependency not from the original location. 
For example here is my original get_db dependency location
`from app.database import get_db`

Before that I tried to import it from app.main or app.dependency and it failed

I would be very grateful if someone explains, why this actually happens! 🙏🏼

**Reply by SayeedKhan21** [reply]
hey man,  i haven't really worked on the code since i posted this issue, however i do remember that changing some import statements did affect the way it worked, but don't know why  , would be great if someone would clear it out

**Comment by fabianfalon** [comment]
Can you try something like that?

```
def override_get_user():
    return "asdf"


class TestApi:
    app.dependency_overrides[get_user] = override_get_user
    client = TestClient(app)

class TestHealthcheck(TestApi):
    def test_healthcheck(self):
        response = self.client.get("/healthcheck")
        assert response.status_code == 200

```
